### PR TITLE
Update fixed hash and uint

### PR DESCRIPTION
**Pull request by Robbepop**
Major overhaul for `ethereum-types` and minor one for `ethbloom` crates.

Both got updated dependencies:
- `fixed-hash 0.2` => `fixed-hash 0.3.0`
- `uint 0.4` => `uint 0.5.0`
- `crunchy 0.1.6` => `crunchy 0.2.1`

Many hash and uint types have been removed from `ethereum-types` as well as some functionality and formerly deprecated items.

This should also lower compilation times of `ethereum-types` a lot.

TODO:
- [ ] Use the `ethereum-types` update in `parity-ethereum` (huge list of dependencies)
- [ ] Use the `ethereum-types`  update in `jsonrpc`
Also `uint` dependency no longer automatically generates its `U256` and `U512` types.

### CODE REVIEW

File: ethereum-types/src/hash.rs
Comment (dvdplm) [inline ethereum-types/src/hash.rs]: Maybe we can remove these `deprecated` methods now?

File: ethereum-types/src/hash.rs
Comment (dvdplm) [inline ethereum-types/src/hash.rs]: This test (and the one after) is kind of weird, it's testing hash features. I think we should remove them or move them to `fixed-hash` (unless there are already tests for this there, and I think there are).

File: ethbloom/Cargo.toml
Comment (dvdplm) [inline ethbloom/Cargo.toml]: Do we really need this? :/

File: ethereum-types/src/hash.rs
Comment (dvdplm) [inline ethereum-types/src/hash.rs:71]: I know this is a lot to ask for, but I for one would really appreciate a rough idea of what we're using all of these hash types for. Are they all in use? Can we deprecate some? Not blocking this PR of course, just something worth pondering.

File: ethereum-types/src/lib.rs
Comment (dvdplm) [inline ethereum-types/src/lib.rs]: It's baaaaaack! AAAARGHHH! 🥶

Can we deprecate this please?

**Review by dvdplm — CHANGES_REQUESTED** [review]

File: ethereum-types/src/hash.rs
Comment (Robbepop) [reply ethereum-types/src/hash.rs]: yup! good idea

File: ethereum-types/src/hash.rs
Comment (Robbepop) [reply ethereum-types/src/hash.rs:71]: With the new `construct_fixed_hash` macro we are able to mark them some as deprecated. :)

File: ethereum-types/src/lib.rs
Comment (Robbepop) [reply ethereum-types/src/lib.rs]: hahaha, yeah I look into it!

File: ethereum-types/src/hash.rs
Comment (Robbepop) [reply ethereum-types/src/hash.rs]: Please do note that I wanted the changes in this crate to be minimal to the outside ecosystem so that I can refocus on other work as soon as possible instead of fixing all the to-be-broken depending crates. That's the reason I did not change some of the things that should have been done otherwise such as renaming the crate features etc.

Comment (Robbepop) [reply ethereum-types/src/hash.rs]: yeah we should do this! you are right

File: ethereum-types/src/hash.rs
Comment (Robbepop) [reply ethereum-types/src/hash.rs]: Please do note that I wanted the changes in this crate to be minimal to the outside ecosystem so that I can refocus on other work as soon as possible instead of fixing all the to-be-broken depending crates. That's the reason I did not change some of the things that should have been done otherwise such as renaming the crate features etc.

File: ethbloom/Cargo.toml
Comment (Robbepop) [reply ethbloom/Cargo.toml]: I don't know atm. I will look into it but wanted to keep the breakage of dependent crates as small as possible to be able to refocus my work as soon as possible.

Comment (dvdplm) [reply ethbloom/Cargo.toml]: But `rustc-hex` was added in this PR, yes?

**Comment by Robbepop** [comment]
I completely revamped this PR.
It is now a lot more aggressive.
Removes many hash and uint types that are not needed.

Note: Deprecation of those items was not feasible since of the never ending warning stream resulting from it.

File: ethereum-types/Cargo.toml
Comment (dvdplm) [inline ethereum-types/Cargo.toml:10]: `version = "0.2"…`

File: ethereum-types/Cargo.toml
Comment (dvdplm) [inline ethereum-types/Cargo.toml]: `version = "0.3"`

File: ethereum-types/Cargo.toml
Comment (dvdplm) [inline ethereum-types/Cargo.toml]: `version = "0.5"`

File: ethereum-types/src/hash.rs
Comment (dvdplm) [inline ethereum-types/src/hash.rs]: `parity-ethereum` uses both `U64` and `U128` currently. Do you have a plan for replacing them?

File: ethereum-types/src/hash.rs
Comment (dvdplm) [inline ethereum-types/src/hash.rs]: Looks like `parity-ethereum` uses all of the three above.

File: ethereum-types/src/hash.rs
Comment (dvdplm) [inline ethereum-types/src/hash.rs]: Used in `parity-ethereum`

**Review by dvdplm — COMMENTED** [review]
Before merging I'd like to see that `parity-ethereum` (and perhaps `jsonrpc`?) works with the new version. Maybe you tried that already though?

File: ethereum-types/src/hash.rs
Comment (Robbepop) [reply ethereum-types/src/hash.rs]: I checked some of the dependent crates on crates.io and couldn't find a crate there that uses them - but maybe just bad luck. To me the usecase for having a `U64` is interesting. What does it bring to the table what a bare metal `u64` doesn't?

I did not plan to invest much more time into refactoring dependent crates.
Maybe we should revert these removals again if they lead to disadvantages.

File: ethereum-types/src/hash.rs
Comment (Robbepop) [reply ethereum-types/src/hash.rs]: see my comment above

File: ethereum-types/src/hash.rs
Comment (Robbepop) [reply ethereum-types/src/hash.rs]: So we actually can't get rid of any of those hash types.
I will revert the removals to make this no longer break things so that we can merge it.

File: ethbloom/Cargo.toml
Comment (Robbepop) [reply ethbloom/Cargo.toml]: Yes it was added. I will investigate.

File: ethereum-types/src/hash.rs
Comment (Robbepop) [reply ethereum-types/src/hash.rs:71]: The problem is that deprecating is not a viable option unless we want to have our terminals full of warnings. Even deprecating only the `H32` was hilarious.

File: ethereum-types/Cargo.toml
Comment (Robbepop) [reply ethereum-types/Cargo.toml:10]: There is a problem. `crunchy` `2.0` and `crunchy` `2.1` are totally different in respect to their default features. So I would keep that one to make our use case explicit.

File: ethereum-types/src/hash.rs
Comment (Robbepop) [reply ethereum-types/src/hash.rs]: removed deprecated methods and impls

File: ethereum-types/src/hash.rs
Comment (Robbepop) [reply ethereum-types/src/hash.rs]: done

File: ethereum-types/src/hash.rs
Comment (Robbepop) [reply ethereum-types/src/hash.rs:71]: tried to remove many. but reverted the change now. so we should do this in another PR.

File: ethereum-types/src/lib.rs
Comment (Robbepop) [reply ethereum-types/src/lib.rs]: removed - just not shown in github somehow

**Review by lexfrl — APPROVED** [review]
lgtm

File: ethereum-types/src/hash.rs
Comment (Robbepop) [reply ethereum-types/src/hash.rs]: reverted removals

File: ethereum-types/Cargo.toml
Comment (Robbepop) [reply ethereum-types/Cargo.toml]: done

File: ethereum-types/Cargo.toml
Comment (Robbepop) [reply ethereum-types/Cargo.toml]: done

**Comment by Robbepop** [comment]
> Before merging I'd like to see that parity-ethereum (and perhaps jsonrpc?) works with the new version. Maybe you tried that already though?

Haven't tried yet but will do.

**Comment by Robbepop** [comment]
I think checking if `parity-ethereum` works with the `ethereum-types` update isn't simply done on a day or so ...
```
cargo tree | grep ethereum-types
```
Shows us this:
```
│   │   ├── ethereum-types-serialize v0.2.1
│   ├── ethereum-types v0.4.0
│   │   ├── ethereum-types-serialize v0.2.1 (*)
│       ├── ethereum-types v0.4.0 (*)
│       │   ├── ethereum-types v0.4.0 (*)
│       │   ├── ethereum-types v0.4.0 (*)
│       │       ├── ethereum-types v0.4.0 (*)
│   │   ├── ethereum-types v0.4.0 (*)
│   │   │   ├── ethereum-types v0.4.0 (*)
│   │   │   ├── ethereum-types v0.4.0 (*)
│   │   ├── ethereum-types v0.4.0 (*)
│   │   ├── ethereum-types v0.4.0 (*)
│   │   │   ├── ethereum-types v0.4.0 (*)
│   │   │   │   ├── ethereum-types v0.4.0 (*)
│   │   │   │   ├── ethereum-types v0.4.0 (*)
│   │   │   │       ├── ethereum-types v0.4.0 (*)
│   │   │   │       │   ├── ethereum-types v0.4.0 (*)
│   │   ├── ethereum-types v0.4.0 (*)
│   │   ├── ethereum-types v0.4.0 (*)
│   ├── ethereum-types v0.4.0 (*)
│   │   ├── ethereum-types v0.4.0 (*)
│   │   ├── ethereum-types v0.4.0 (*)
│   │   └── ethereum-types v0.4.0 (*)
│   │   ├── ethereum-types v0.4.0 (*)
│       ├── ethereum-types v0.4.0 (*)
│   │   ├── ethereum-types v0.4.0 (*)
│   ├── ethereum-types v0.4.0 (*)
│   ├── ethereum-types v0.4.0 (*)
│   │   │   ├── ethereum-types v0.4.0 (*)
│   │   ├── ethereum-types v0.4.0 (*)
│   ├── ethereum-types v0.4.0 (*)
├── ethereum-types v0.4.0 (*)
│   ├── ethereum-types v0.4.0 (*)
│   ├── ethereum-types v0.4.0 (*)
│   ├── ethereum-types v0.4.0 (*)
│   │   ├── ethereum-types v0.4.0 (*)
│   ├── ethereum-types v0.4.0 (*)
│   │   ├── ethereum-types v0.4.0 (*)
│   ├── ethereum-types v0.4.0 (*)
```

**Comment by Robbepop** [comment]
I added a TODO list for this PR.

File: ethbloom/Cargo.toml
Comment (sorpaas) [inline ethbloom/Cargo.toml]: I think you can just remove this line. This dep is not used at all in code.

**Review by sorpaas — APPROVED** [review]
LGTM. Left one final grumble.

@dvdplm Maybe we can merge this? If we want to integrate this to `parity-ethereum` and `jsonrpc`, we'd need to first merge and publish new versions anyway.

**Comment by Robbepop** [comment]
> LGTM. Left one final grumble.
>
> @dvdplm Maybe we can merge this? If we want to integrate this to parity-ethereum and jsonrpc, we'd need to first merge and publish new versions anyway.

@sorpaas  We did not merge this yet because we were missing a person that wanted to take over updating of the remaining dependencies which @dvdplm suspects to be a lot. I will fix the grumble asap.

File: ethbloom/Cargo.toml
Comment (Robbepop) [reply ethbloom/Cargo.toml]: done

**Comment by Robbepop** [comment]
We should release a version `0.5` of `ethereum-types` after merging this.

**Comment by sorpaas** [comment]
@Robbepop @dvdplm Thanks! If no objections, let's merge this?

**Comment by Robbepop** [comment]
> @Robbepop @dvdplm Thanks! If no objections, let's merge this?

You are free to merge! :)

**Pull request merged**
